server/src/main/java/org/elasticsearch/transport/netty4/AdaptiveSslHandler.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package org.elasticsearch.transport.netty4;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.ssl.OptionalSslHandler;
+import io.netty.handler.ssl.SslHandler;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLException;
+import java.net.InetSocketAddress;
+import java.util.List;
+
+/**
+ * This class is hybrid of {@link SslHandler} and {@link OptionalSslHandler}
+ * It behaves like SslHandler for outgoing messages and encrypted incoming messages
+ * and behaves like OptionalSslHandler for unencrypted incoming messages.
+ */
+public class AdaptiveSslHandler extends SslHandler {
+
+    private static final Logger LOGGER = LogManager.getLogger(AdaptiveSslHandler.class);
+    private static final int SSL_RECORD_HEADER_LENGTH = 5;
+
+    public AdaptiveSslHandler(SSLEngine engine) {
+        super(engine);
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws SSLException {
+        // Inspired by OptionalSslHandler.decode.
+
+        if (in.readableBytes() < SSL_RECORD_HEADER_LENGTH) {
+            return;
+        }
+
+        if (SslHandler.isEncrypted(in)) {
```

Comment:
Is it expensive to run this check? If it is we could consider to replace the `AdaptiveSslHandler` with the regular handler after some point, and only allow the downgrade happening after the initial handshake

Reply:
It analyses only first 5 bytes - internally it calls Netty's `getEncryptedPacketLength` which is basically same with open search check ([our adaptation](https://github.com/crate/crate/commit/a7416f9861a12a0a8c8bf9ac57c2bbbf69f843da#diff-1c5d1afa88e5576595f2438652d0bc5bd960477875add1f6e0b6a06341765de1R103-R104))
Anyway, if can improve, let's go for it - I can create a follow up PR to keep this one small if it's fine

Reply:
Okay - sounds like it isn't worth optimizing